add tidymodels 2023 q4 post (#677)
tidyverse/tidyverse.org: content/blog/tidymodels-2023-q4/index.Rmd

-   [Three ways errors are about to get better in tidymodels](https://www.tidyverse.org/blog/2023/11/tidymodels-errors-q4/)

Since [our last roundup post](https://www.tidyverse.org/blog/2022/12/tidymodels-2022-q4/), there have been CRAN releases of `r nrow(tm_data)` tidymodels packages. Here are links to their NEWS files:

```{r}
#| echo: FALSE
#| results: asis

cat(tm_data$news, sep = "\n")
```

We'll highlight a few especially notable changes below: Updated warnings when normalizing, and better error messages in recipes.

```{r}
#| message: FALSE
library(tidymodels)

data("ames", package = "modeldata")
```

## Updated warnings when normalizing

The latest release of recipes features an overhaul of the warnings and error messages to use the [cli](https://cli.r-lib.org/) package. With this, we are starting the project of providing more information signaling when things don't go well.

> The first type of issue we now signal for is when you try to normalize data that contains non-numeric numbers such as `NA` or `Inf`. These can sneak in for several reasons, and before this release, it happened silently. Below we are creating a recipe using the `ames` data set, and before we normalize we are taking the logarithms root of all variables that pertain to square footage. 

Reviewer: ```suggestion
The first type of issue we now signal for is when you try to normalize data that contains elements such as `NA` or `Inf`. These can sneak in for several reasons, and before this release, it happened silently. Below we are creating a recipe using the `ames` data set, and before we normalize, we are taking the logarithms root of all variables that pertain to square footage. 
```

```{r}
rec <- recipe(Sale_Price ~ ., data = ames) |>
  step_log(contains("SF")) |>
  step_normalize(all_numeric_predictors()) |>
  prep()
```

> We now get a warning that something happened. Telling us that it encountered `Inf` or `-Inf`. Knowing that we can go back and investigate what went wrong. If we exclude `step_normalize()` and `bake()` the recipe, we see that a number of `-Inf` appears

Reviewer: ```suggestion
We now get a warning that something happened. Telling us that it encountered `Inf` or `-Inf`. Knowing that we can go back and investigate what went wrong. If we exclude `step_normalize()` and `bake()` the recipe, we see that a number of `-Inf` appears.
```

---
output: hugodown::hugo_document

slug: tidymodels-2023-q4
title: "Q4 2023 tidymodels digest"
date: 2024-01-08
author: Emil Hvitfeldt
description: >
    The tidymodels team has been busy working on all sorts of new features 
    across the ecosystem.
photo:

>   url: https://www.pexels.com/photo/landscape-photography-of-snow-pathway-between-trees-during-winter-688660/

Reviewer 2: Love the picture. :)

Author: thank you! it was the first thing that popped up when i searched for "winter" 😆  https://www.pexels.com/search/winter/

  )

txt_pkg_list <- knitr::combine_words(tm_data$package)
```

The [tidymodels](https://www.tidymodels.org/) framework is a collection of R packages for modeling and machine learning using tidyverse principles.

Since the beginning of 2021, we have been publishing [quarterly updates](https://www.tidyverse.org/categories/roundup/) here on the tidyverse blog summarizing what's new in the tidymodels ecosystem. The purpose of these regular posts is to share useful new features and any updates you may have missed. You can check out the [`tidymodels` tag](https://www.tidyverse.org/tags/tidymodels/) to find all tidymodels blog posts here, including our roundup posts as well as those that are more focused, like this post from the past couple of months:

-   [Three ways errors are about to get better in tidymodels](https://www.tidyverse.org/blog/2023/11/tidymodels-errors-q4/)

Since [our last roundup post](https://www.tidyverse.org/blog/2022/12/tidymodels-2022-q4/), there have been CRAN releases of `r nrow(tm_data)` tidymodels packages. Here are links to their NEWS files:

```{r}
#| echo: FALSE
#| results: asis

cat(tm_data$news, sep = "\n")
```

> We'll highlight a few especially notable changes below: Updated warnings when normalizing, and better error messages in recipes.

Reviewer 2: ```suggestion
We'll highlight a few especially notable changes below: updated warnings when normalizing, and better error messages in recipes.
```

```{r}
#| message: FALSE
library(tidymodels)

data("ames", package = "modeldata")
```

## Updated warnings when normalizing

The latest release of recipes features an overhaul of the warnings and error messages to use the [cli](https://cli.r-lib.org/) package. With this, we are starting the project of providing more information signaling when things don't go well.

> The first type of issue we now signal for is when you try to normalize data that contains non-numeric numbers such as `NA` or `Inf`. These can sneak in for several reasons, and before this release, it happened silently. Below we are creating a recipe using the `ames` data set, and before we normalize we are taking the logarithms root of all variables that pertain to square footage. 

Reviewer 2: ```suggestion
The first type of issue we now signal for is when you try to normalize data that contains non-numeric numbers such as `NA` or `Inf`. These can sneak in for several reasons, and before this release, it happened silently. Below we are creating a recipe using the `ames` data set, and before we normalize we are taking the logarithms of all variables that pertain to square footage. 
```

```{r}
rec <- recipe(Sale_Price ~ ., data = ames) |>
  step_log(contains("SF")) |>
  step_normalize(all_numeric_predictors()) |>
  prep()
```

> We now get a warning that something happened. Telling us that it encountered `Inf` or `-Inf`. Knowing that we can go back and investigate what went wrong. If we exclude `step_normalize()` and `bake()` the recipe, we see that a number of `-Inf` appears

Reviewer 2: ```suggestion
We now get a warning that something happened, telling us that it encountered `Inf` or `-Inf`. Knowing that, we can go back and investigate what went wrong. If we exclude `step_normalize()` and `bake()` the recipe, we see that a number of `-Inf` values appear
```

```{r}
recipe(Sale_Price ~ ., data = ames) |>
  step_log(contains("SF")) |>
  prep() |>
  bake(new_data = NULL, contains("SF")) |>
  glimpse()
```

> Looking at the bare data set, we notice that the `-Inf` all appear where there are `0`, which makes sense since `log(0)`.

Reviewer 2: ```suggestion
Looking at the bare data set, we notice that the `-Inf` all appear where there are `0`, which makes sense since `log(0)` is undefined.
```

Knowing that it was `0` that caused the problem, we can set an `offset` to avoid taking `log(0)`.

```{r}
rec <- recipe(Sale_Price ~ ., data = ames) |>
  step_log(contains("SF"), offset = 0.5) |>
  step_normalize(all_numeric_predictors()) |>
  prep()
```

These warnings appear in `step_scale()`, `step_normalize()`, `step_center()` or `step_range()`.

## Better error messages in recipes

Another problem that happens a lot when using recipes, is accidentally selecting variables that have the wrong types. Previously this caused the following error:

```{r}
#| eval: false
recipe(Sale_Price ~ ., data = ames) |>
  step_dummy(starts_with("Lot_")) |>
  prep()
#> Error in `step_dummy()`:
#> Caused by error in `prep()`:
#> ! All columns selected for the step should be string, factor, or ordered.
```

In the newest release, it will detail the offending variables and what was wrong with them.

```{r}
#| error: true
recipe(Sale_Price ~ ., data = ames) |>
  step_dummy(starts_with("Lot_")) |>
  prep() |>
  bake()
```

## Coming Attractions

In the next month or so we are planning a cascade of CRAN releases. There is a lot of new functionality coming your way, especially in the tune package. 

> A number of our packages will (finally) be able to cohesively fit, evaluate, tune, and predict models for event times (a.k.a., [survival analysis](https://en.wikipedia.org/wiki/Survival_analysis)). If you don't do this type of work, you might not notice the new capabilities. However, if you do, tidymodels will be able to do a lot more for you. 

Reviewer 2: ```suggestion
A number of our packages will (finally) be able to cohesively fit, evaluate, tune, and predict models for event times (a.k.a., [survival analysis](https://en.wikipedia.org/wiki/Survival_analysis)). If you don't do this type of work, you might not notice the new capabilities. However, if you do, tidymodels will be able to do a lot more for you. 

We've also implemented a number of features related to model fairness. These tools allow tidymodels users to identify when machine learning models behave unfairly towards certain groups of people, and will also be included in the upcoming releases of tidymodels packages in Q1.
```

Reviewer 2: I understand that this wasn't as much of a technical feat as survival analysis, but seems worth calling out if we're referring to coming attractions. :)

Reviewer 3: Sorry., I should have added that. I've got survival fever.

> In addition to the new packages, we’ll highlight a lot of these new capabilities in blog posts here as well as tutorials on [`tidymodels.org`](https://www.tidymodels.org/). 

Reviewer: ```suggestion
We’ll highlight a lot of these new capabilities in blog posts here as well as tutorials on [`tidymodels.org`](https://www.tidymodels.org/). 
```
